File: kafka-consumer/src/main/java/net/opentsdb/kafka/consumer/TsdbConsumer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package net.opentsdb.kafka.consumer;

import com.google.common.util.concurrent.AbstractIdleService;
import com.google.inject.Inject;
import com.google.inject.Singleton;
import com.google.inject.name.Named;
import kafka.consumer.*;
import kafka.consumer.Consumer;
import kafka.javaapi.consumer.ConsumerConnector;
import kafka.message.Message;
import net.opentsdb.client.netty.TsdbClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.CharsetDecoder;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

/**
 * Kafka Consumer pulls the data out of kafka and sends it to the OpenTSDB daemon
 */
@Singleton
public class TsdbConsumer extends AbstractIdleService {
  private static final String KAFKA_TOPIC = "tsdb.kafka.topic";

  private static final Logger logger = LoggerFactory.getLogger(TsdbConsumer.class);

  private static final Charset charset = Charset.forName("UTF-8");
  private static final CharsetDecoder decoder = charset.newDecoder();

  private final ConsumerConnector consumerConnector;
  private final TsdbClient client;
  private final String topic;

  @Inject
  public TsdbConsumer(TsdbClient client, Properties properties, @Named(KAFKA_TOPIC) String topic) {
    this.client = client;
    this.topic = topic;

    ConsumerConfig config = new ConsumerConfig(properties);

    this.consumerConnector = Consumer.createJavaConsumerConnector(config);
  }

  @Override
  protected void startUp() throws Exception {
    logger.info("Starting up TSDB Kafka Consumer");
    Map<String, Integer> topicCountMap = new HashMap<>();
    topicCountMap.put(topic, 1);
    Map<String, List<KafkaStream<Message>>> consumerMap = consumerConnector.createMessageStreams(topicCountMap);
    KafkaStream<Message> stream =  consumerMap.get(topic).get(0);
    ConsumerIterator<Message> it = stream.iterator();
    while(it.hasNext()) {
      Message message = it.next().message();
      ByteBuffer buffer = message.payload();
      byte[] bytes = new byte[buffer.remaining()];
      buffer.get(bytes);
      String messageString = new String(bytes);
      logger.debug("Message is: {}", messageString);
      client.send(messageString);
    }
  }

  @Override
  protected void shutDown() throws Exception {
    logger.info("Shutting down TSDB Kafka Consumer");
    consumerConnector.shutdown();
  }
}

File: kafka-consumer/src/main/scala/net/opentsdb/kafka/consumer/TsdbConsumer.scala
package net.opentsdb.kafka.consumer

import com.google.inject.Inject
import com.google.common.util.concurrent.AbstractIdleService
import java.util.Properties
import com.google.inject.name.Named
import org.slf4j.LoggerFactory
import kafka.consumer.{ConsumerConfig, Consumer}
import java.util.concurrent.Executors
import javax.inject.Singleton
import kafka.serializer.{DefaultDecoder, StringDecoder}
import net.opentsdb.core.{Tags, TSDB}
import net.opentsdb.stats.StatsCollector
import java.util

@Singleton
class TsdbConsumer @Inject() (props: Properties, tsdbClient: TSDB, @Named("tsdb.kafka.topic") topicName: String,
                              @Named("tsdb.kafka.topic.partitions") partitions: Int) extends AbstractIdleService{
  private final val logger = LoggerFactory.getLogger(classOf[TsdbConsumer])
  private final val consumerConnector = Consumer.create(new ConsumerConfig(props))
  private final val executorPool = Executors.newFixedThreadPool(partitions, new ConsumerThreadFactory)


  def startUp() {

    tsdbClient.collectStats(new StatsCollector("tsd") {
      def emit(line: String) = logger.info(line)
    })

    logger.info("Starting up TSDB Kafka Consumer")
    val topicMessageStreams = consumerConnector.createMessageStreams(Predef.Map(topicName -> partitions), new DefaultDecoder,
      new StringDecoder())
    val topicStreams = topicMessageStreams.get(topicName).get

    for(stream <- topicStreams) {
      executorPool.execute(new Runnable() {
        override def run() {
          for(kafkaMessage <- stream) {
            val words = Tags.splitString(kafkaMessage.message, ' ')

            val metric = words{1}
            val timestamp = Tags.parseLong(words{2})
            val value = words{3}

            val tags: util.HashMap[String, String] = new util.HashMap[String, String]

            for (i <- 4 until words.length if !words(i).isEmpty) {
              Tags.parse(tags, words(i))
            }

            if (value.equals("NaN") || value.startsWith("N")) {
              tsdbClient.addPoint(metric, timestamp, 0, tags)
            } else if (Tags.looksLikeInteger(value)) {
              tsdbClient.addPoint(metric, timestamp, Tags.parseLong(value), tags)
            } else {
              tsdbClient.addPoint(metric, timestamp, java.lang.Float.parseFloat(value), tags)
            }
          }
        }
      })
    }
  }

  def shutDown() {
    logger.info("Shutting down TSDB Kafka Consumer")
    tsdbClient.shutdown().joinUninterruptibly()
    executorPool.shutdown()
    consumerConnector.shutdown()
  }
}


